### ISSUE TITLE
Unexpected Post Hoc Results

### ISSUE THREAD

**Issue by joshuatbland** [0todd0000/spm1d#293]
Hi. Hello. I have run a one-way repeated measures ANOVA on a dataset - I have included a plot of the ANOVA results and a plot for each subject. The ANOVA results suggest that there are significant within-subject differences between conditions along the length of the trajectory. However, post hoc tests do not show any significant results - this is very surprising. Any insights into what might be causing this would be greatly appreciated. 
![Unexpected Post-Hoc Results](https://github.com/0todd0000/spm1d/assets/137851275/e07e3915-1f1a-4aba-976d-fc5fbf43cdc0)

**Comment by 0todd0000** [comment]
If you have used `spm1d.stats.ttest2` for *post hoc* tests you may indeed see different results. If you use `spm1d.stats.ttest_paired` you should see *post hoc* results that more closely agree with these one-way RM-ANOVA results.

**Comment by joshuatbland** [comment]
We did indeed use the paired t-test for our post hoc tests but we did not get results that agree with the one-way RM-ANOVA results. Here our the results for the paired t-tests ( with Bonferroni correction):
![Pired_t_test_results](https://github.com/0todd0000/spm1d/assets/137851275/85080d9f-5624-40cf-91f8-79275c76af33)

**Comment by 0todd0000** [comment]
Everything you've done sounds find, and I agree that these *post hoc* results are unexpected given the ANOVA results. To help me understand the analyses more completely can you please attach or copy-paste a script that generates the figures above?

**Comment by joshuatbland** [comment]
Yes, here is the code that was used to generate the figures: 

```
%% One-Way Repeated Measures Anova 
%%--------------------------------------------------------------
[Y,A,SUBJ] = deal(ang,condition, specimen);

spm = spm1d.stats.anova1rm(Y, A, SUBJ);
spmi = spm.inference(0.05);

spmi.plot();
spmi.plot_threshold_label();
spmi.plot_p_values();

%% T-test Post-Hoc Tests 
%%--------------------------------------------------------------
alpha = 0.05
nTests = 3
p_critical = spm1d.util.p_critical_bonf(alpha, nTests)

% Condition 1 vs Condition 2
figure()
spm_one = spm1d.stats.ttest_paired(ang(cond1_indexes,:), ang(cond2_indexes,:));
spmi_one = spm_one.inference(p_critical, 'two_tailed', true, 'interp',true);
spmi_one.plot();

% Condition 1 vs Condition 3 
figure()
spm_two = spm1d.stats.ttest_paired(ang(cond1_indexes,:), ang(cond3_indexes,:));
spmi_two = spm_two.inference(p_critical, 'two_tailed', true, 'interp',true);
spmi_two.plot();

% Condition 2 vs Condition 3
figure()
spm_three = spm1d.stats.ttest_paired(ang(cond2_indexes,:), ang(cond3_indexes,:));
spmi_three = spm_three.inference(p_critical, 'two_tailed', true, 'interp',true);
spmi_three.plot();
```

**Comment by 0todd0000** [comment]
Everything looks fine in the script. I have two follow-up questions:


Can you please confirm that the index vectors are defined as follows?

```octave
cond1_indexes = A==1;
cond2_indexes = A==2;
cond3_indexes = A==3;
```

Can you please confirm that the following three vectors are identical?

```octave
subj1 = SUBJ(cond1_indexes);
subj2 = SUBJ(cond2_indexes);
subj3 = SUBJ(cond3_indexes);
```

**Comment by joshuatbland** [comment]
Yes, I was able to confirm both of these.

**Comment by 0todd0000** [comment]
OK, thanks. Can you please tell me:

- the number of subjects
- the shape of the `Y` array

**Comment by joshuatbland** [comment]
- Number of subjects = 6
- Shape of Y array = 6 by 828

**Comment by 0todd0000** [comment]
OK, thank you for these details. There appears to have been a misorganization of `Y`.

Each row of `Y` should be a single 1D observation with `Q` points. Usually `Q=100` or `Q=101`.

Since there are six subjects and three conditions, and assuming there is one observation per subject/condition pair, then the shape of `Y` should be: `(18,Q)`.  And `A` and `SUBJ` should look something like this:

```octave
A    = [1, 1, 1, 1, 1, 1,    2, 2, 2, 2, 2, 2,    3, 3, 3, 3, 3, 3]';
SUBJ = [1, 2, 3, 4, 5, 6,    1, 2, 3, 4, 5, 6,    1, 2, 3, 4, 5, 6]';
```

**Comment by joshuatbland** [comment]
Oh, sorry - I was looking at the shape of the wrong variable. The shape of the Y array is actually 18 by 828 (Q of 828).

**Comment by joshuatbland** [comment]
Is it okay to have Q be 828 as long as all of the observations have 828 points?

**Comment by 0todd0000** [comment]
`Q=828` is no problem.

However, I am confused: why do each of the individual subject figures in your original post show `Q` closer to 100?

Also, can you please send the values of `A` and `SUBJ` and/or verify that your `A` and `SUBJ` values are as follows?

```octave
A    = [1, 1, 1, 1, 1, 1,    2, 2, 2, 2, 2, 2,    3, 3, 3, 3, 3, 3]';
SUBJ = [1, 2, 3, 4, 5, 6,    1, 2, 3, 4, 5, 6,    1, 2, 3, 4, 5, 6]';
```

**Comment by joshuatbland** [comment]
The individual plots show the output value plotted against another value (flexion angle) but there are still 828 points. Yes, those are the values of A and SUBJ.

**Comment by 0todd0000** [comment]
OK, thank you for confirming. Based on your descriptions I am unable to explain the *post hoc* results. I think the only way to proceed would be for me to look at the actual data to see if I can find out what might be going on. Can you please send your data file(s) and .m script to this address?  consulting@spm1d.org

**Comment by joshuatbland** [comment]
Yes, I just sent the files to the email address. Thank you so much. Let me know if I can provide any additional information.

**Comment by 0todd0000** [comment]
Thank you for sending your data and apologies for the delay. 

I've discovered the source of the problem: non-normality.

Consider the figure below:
- (a) Parametric inference results (`spm1d.stats.anova1rm`)
- (b) **Nonparametric** inference results (`spm1d.stats.nonparam.anova1rm`)
- (c) Groups 1 & 2 observations and means
- (d) *Post hoc* parametric paired comparison of Groups 1 & 2 (uncorrected; `spm1d.stats.ttest_paired`)
- (e) *Post hoc* **nonparametric** paired comparison of Groups 1 & 2 (uncorrected; `spm1d.stats.nonparam.ttest_paired`)
- (f) Paired differences and the mean paired difference for Groups 1 & 2

The key panels are (e) and (f).  From panel (f) it is clear that these data are non-normal, with just one observation near the mean and most observations far away from the mean. This appears to be a bimodal distribution, and since it is non-normal it is more appropriate to use nonparametric inference.

If nonparametric inference is used, as in panel (e), then the *post hoc* results (e) agree much more closely with the main ANOVA results (b).



![fig](https://github.com/user-attachments/assets/333fe222-a787-4dd3-bf9e-1956fad3d3c7)

**Comment by joshuatbland** [comment]
Okay, great. Thanks for taking a look at this data - this is very helpful!

**Comment by joshuatbland** [comment]
Unfortunately, I was unable to to produce the same output for the paired t-tests. Am I setting something up wrong. I have attached my code and output plots below: 

`clear;  clc

% load the data
load('data.mat');

%% One-Way Repeated Measures Anova 
%%--------------------------------------------------------------
rng(0)
[Y,A,SUBJ] = deal(ang,condition, specimen);

spm = spm1d.stats.nonparam.anova1rm(Y, A, SUBJ);
spmi = spm.inference(0.05,'iterations',100);

spmi.plot();
spmi.plot_threshold_label();
spmi.plot_p_values();
title('One-Way Repeated Measures Anova');

%% T-test Post-Hoc Tests 
%%--------------------------------------------------------------
alpha = 0.05
nTests = 3
p_critical = spm1d.util.p_critical_bonf(alpha, nTests)

% Condition 1 vs Condition 2
figure()
rng(0)
spm_one = spm1d.stats.nonparam.ttest_paired(ang(cond1_indexes,:), ang(cond2_indexes,:));
spmi_one = spm_one.inference(p_critical, 'two_tailed', true, 'interp',true, 'iterations',-1);
spmi_one.plot();
title('Condition 1 vs Condition 2');

% Condition 1 vs Condition 3 
figure()
rng(0)
spm_two = spm1d.stats.nonparam.ttest_paired(ang(cond1_indexes,:), ang(cond3_indexes,:));
spmi_two = spm_two.inference(p_critical, 'two_tailed', true, 'interp',true, 'iterations',-1);
spmi_two.plot();
title('Condition 1 vs Condition 3');

% Condition 2 vs Condition 3
figure()
rng(0)
spm_three = spm1d.stats.nonparam.ttest_paired(ang(cond2_indexes,:), ang(cond3_indexes,:));
spmi_three = spm_three.inference(p_critical, 'two_tailed', true, 'interp',true, 'iterations',-1);
spmi_three.plot();
title('Condition 2 vs Condition 3');
`
![Presentation1](https://github.com/user-attachments/assets/4c9f4dd3-f957-4e26-9f10-cc8a474b4111)

**Comment by 0todd0000** [comment]
It looks like my results are the same as your Condition 2 vs. 3 results. I used an uncorrected threshold for illustrative purposes.
